relay-server/src/actors/healthcheck.rs

```diff
@@ -17,16 +17,33 @@ use crate::statsd::RelayGauges;
 
 lazy_static::lazy_static! {
     /// Singleton of the `Healthcheck` service.
-    static ref ADDRESS: RwLock<Option<Addr<HealthcheckMessage>>> = RwLock::new(None);
+    static ref ADDRESS: RwLock<Option<Addr<Healthcheck>>> = RwLock::new(None);
 }
 
-/// Internal wrapper of a message sent through an `Addr` with return channel.
-#[derive(Debug)]
-struct Message<T> {
-    data: T,
-    // TODO(tobias): This is hard-coded to return `bool`.
-    // Might need some associated types to make this work
-    responder: oneshot::Sender<bool>,
+/// Our definition of a service.
+///
+/// Services are much like actors: they receive messages from an inbox and handles them one
+/// by one.  Services are free to concurrently process these messages or not, most probably
+/// should.
+///
+/// Messages always have a response which will be sent once the message is handled by the
+/// service.
+pub trait Service {
+    /// The envelope is what is sent to the inbox of this service.
+    ///
+    /// It is an enum of all the message types that can be handled by this service together
+    /// with the response [sender](oneshot::Sender) for each message.
+    type Envelope;
+}
+
+/// A message which can be sent to a service.
+///
+/// Messages have an associated `Response` type and can be unconditionally converted into
+/// the envelope type of their [`Service`].
+pub trait ServiceMessage<S: Service> {
+    type Response;
+
+    fn into_envelope(self) -> (S::Envelope, oneshot::Receiver<Self::Response>);
```

Comment:
We will have a good amount of messages that are fire-and-forget, so they do not need a oneshot receiver. Could we make the receiver optional?

The safest way would even be to type this as associated type in the service message, so that we can enforce two kinds of `Addr::send` methods via the type system.

Reply:
I'm not sure how to approach this with another associated type, the simplest would be to create an entire new trait like "OnewayServiceMessage" or so.  Though I'm also not really sure how much we care.  We can just let those messages return `()` and not await them (we can abstract this behind `do_send` sure).  That way you still always have the option of checking a message was actually received (I'm no fan of just swallowing the fact that the receiver might be gone, seems like a huge bug and maybe a panic would be better)